In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import re
import tensorflow as tf
from tensorflow import keras
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score
from sklearn.preprocessing import StandardScaler

from IPython.display import display

%matplotlib inline

#first of course we must import the necessary modules

#Boston 311 v3 - Exploring Outliers and Adjusting Data Cleaning Functions

What type of outliers do we have in our data? Let's refer back to the first notebook and take a look at the graphs of our features.

1. subject - 8 of our 10 subject categories have very few caases compared to the other 2
2. reason - Of the more than 40 reason categories, about half have very few records associated with them
3. department - Of our 16 department categories, about half have very few records associated with them
4. source - We have five source categories, but the vast majority of our data is isolated to two of them
5. ward_number - Our ward data is actually fairly normally distributed

Of our five feature categories, only the ward does not have underrepresented categories

For our labels, the logistic label of "Open" or "Closed" shows that the vast majority of cases are eventually closed. Additionally, we don't know if the cases that are open are later closed in the 2023 data set.

Let's count these minority categories and see what's in them:

In [ ]:
df2022 = pd.read_csv("https://data.boston.gov/dataset/8048697b-ad64-4bfc-b090-ee00169f2323/resource/81a7b022-f8fc-4da5-80e4-b160058ca207/download/tmph4izx_fb.csv",
                            parse_dates=['open_dt', 'target_dt', 'closed_dt'])

In [ ]:
subject_counts = df2022['subject'].value_counts()
reason_counts = df2022['reason'].value_counts()
department_counts = df2022['department'].value_counts()
source_counts = df2022['source'].value_counts()

In [ ]:
#minority subjects:
df2022['subject'].value_counts()


Public Works Department              145989
Transportation - Traffic Division     75015
Inspectional Services                 19276
Parks & Recreation Department         16712
Mayor's 24 Hour Hotline               10610
Animal Control                         4027
Property Management                    2902
Boston Water & Sewer Commission        1469
Boston Police Department                701
Neighborhood Services                    22
Name: subject, dtype: int64

In [ ]:
df2022['reason'].value_counts()

Enforcement & Abandoned Vehicles     62656
Street Cleaning                      40582
Code Enforcement                     30933
Sanitation                           29538
Highway Maintenance                  26682
Signs & Signals                      10839
Recycling                             8944
Trees                                 8358
Street Lights                         8224
Park Maintenance & Safety             8052
Housing                               7116
Needle Program                        6845
Building                              6065
Environmental Services                4764
Animal Issues                         4027
Graffiti                              2902
Administrative & General Requests     2075
Employee & General Comments           1835
Health                                1283
Abandoned Bicycle                     1057
Noise Disturbance                      701
Traffic Management & Engineering       646
Notification                           616
Catchbasin 

In [ ]:
df2022['department'].value_counts()

PWDx    136719
BTDT     74856
ISD      18473
PARK     15697
INFO     14643
GEN_      6872
BWSC      4935
PROP      3079
ANML       736
BPD_       298
BHA_       147
BPS_       140
ONS_        79
DND_        27
DISB        19
ECON         3
Name: department, dtype: int64

In [ ]:
df2022['source'].value_counts()

Citizens Connect App    140829
Constituent Call         97507
City Worker App          24618
Self Service              8756
Employee Generated        5013
Name: source, dtype: int64

The graphs were a little deceptive for some of these categories. 

Subjects - The lowest category is neighborhood services, with 22 records, but the next lowest is Boston Police Department, with over 700 records. It doesn't seem like a good idea to remove the BPD records, but we can look at the neighborhood services records:

In [ ]:
from IPython.display import display



display(df2022[df2022['subject'] == "Neighborhood Services"])

case_enquiry_id             open_dt target_dt           closed_dt  \
926        101004126469 2022-01-14 16:42:00       NaT 2022-02-14 11:42:58   
12728      101004290312 2022-05-13 05:49:00       NaT 2022-05-23 17:37:39   
35510      101004148796 2022-01-27 19:51:00       NaT 2022-05-17 17:13:06   
42381      101004123940 2022-01-12 11:53:00       NaT 2022-02-14 11:29:06   
46088      101004116461 2022-01-05 04:52:00       NaT 2022-02-14 11:37:58   
49124      101004164889 2022-02-04 04:53:38       NaT 2022-02-08 12:41:34   
56524      101004163038 2022-02-02 22:14:00       NaT 2022-02-15 14:09:03   
75488      101004227021 2022-03-20 12:35:00       NaT 2022-04-13 10:51:40   
81422      101004257519 2022-04-15 04:24:00       NaT 2022-04-22 18:04:48   
121951     101004169889 2022-02-08 04:54:00       NaT 2022-02-14 13:48:14   
147699     101004270854 2022-04-27 04:11:00       NaT                 NaT   
147953     101004226353 2022-03-19 13:32:00       NaT 2022-04-13 10:46:18   
156599     101004242605 2022-04-03 04:30:00       NaT 2022-04-19 17:33:13   
165328     101004589145 2022-12-02 14:33:00       NaT                 NaT   
168666     101004514936 2022-10-03 17:08:00       NaT 2022-10-18 13:58:21   
193075     101004584731 2022-11-29 06:33:00       NaT 2022-12-01 13:08:38   
210249     101004612130 2022-12-28 06:13:00       NaT 2023-01-05 16:22:39   
212655     101004580633 2022-11-23 10:02:00       NaT                 NaT   
223833     101004457169 2022-08-29 15:28:00       NaT 2022-09-13 15:57:25   
232145     101004579092 2022-11-22 04:17:00       NaT 2022-12-01 13:34:47   
241499     101004610954 2022-12-27 04:44:00       NaT                 NaT   
276291     101004612126 2022-12-28 05:55:00       NaT 2023-01-05 16:24:13   

        ontime case_status                                     closure_reason  \
926     ONTIME      Closed  Case Closed. Closed date : 2022-02-14 11:42:58...   
12728   ONTIME      Closed  Case Closed. Closed date : 2022-05-23 17:37:39...   
35510   ONTIME      Closed  Case Closed. Closed date : 2022-05-17 17:13:06...   
42381   ONTIME      Closed  Case Closed. Closed date : 2022-02-14 11:29:06...   
46088   ONTIME      Closed  Case Closed. Closed date : 2022-02-14 11:37:58...   
49124   ONTIME      Closed  Case Closed. Closed date : 2022-02-08 12:41:34...   
56524   ONTIME      Closed  Case Closed. Closed date : 2022-02-15 14:09:03...   
75488   ONTIME      Closed  Case Closed. Closed date : 2022-04-13 10:51:40...   
81422   ONTIME      Closed  Case Closed. Closed date : 2022-04-22 18:04:48...   
121951  ONTIME      Closed  Case Closed. Closed date : 2022-02-14 13:48:14...   
147699  ONTIME        Open                                                      
147953  ONTIME      Closed  Case Closed. Closed date : 2022-04-13 10:46:18...   
156599  ONTIME      Closed  Case Closed. Closed date : 2022-04-19 17:33:13...   
165328  ONTIME        Open                                                      
168666  ONTIME      Closed  Case Closed. Closed date : 2022-10-18 13:58:21...   
193075  ONTIME      Closed  Case Closed. Closed date : 2022-12-01 13:08:38...   
210249  ONTIME      Closed  Case Closed. Closed date : 2023-01-05 16:22:39...   
212655  ONTIME        Open                                                      
223833  ONTIME      Closed  Case Closed. Closed date : 2022-09-13 15:57:25...   
232145  ONTIME      Closed  Case Closed. Closed date : 2022-12-01 13:34:47...   
241499  ONTIME        Open                                                      
276291  ONTIME      Closed  Case Closed. Closed date : 2023-01-05 16:24:13...   

                                   case_title                subject  \
926     Dumpster & Loading Noise Disturbances  Neighborhood Services   
12728   Dumpster & Loading Noise Disturbances  Neighborhood Services   
35510   Dumpster & Loading Noise Disturbances  Neighborhood Services   
42381   Dumpster & Loading Noise Disturbances  Neighborhood Services   
46088   Dump

These 22 cases are all about Dumpster loading noise complaints related to private trash pickup, and almost all of them mention that according to law the state cannot do anything about private trash pickup times. Interestingly, they all have the same case_title. Is it possible case title is a meaningful and regulated category? Let's count the values:

In [ ]:
len(df2022['case_title'].value_counts().tolist())

6575

With 6575 different values, it probably is not a good categorical variable for us right now.

What about The cases where the reason has less than 25 records?

In [ ]:
reason_filter = df2022['reason'].isin(reason_counts[reason_counts < 25].index)
display(df2022[reason_filter])


case_enquiry_id             open_dt           target_dt  \
926        101004126469 2022-01-14 16:42:00                 NaT   
3165       101004150433 2022-01-29 14:00:00                 NaT   
12728      101004290312 2022-05-13 05:49:00                 NaT   
19321      101004387857 2022-07-13 23:28:00                 NaT   
19608      101004400706 2022-07-24 21:32:00 2022-08-08 08:30:00   
...                 ...                 ...                 ...   
258838     101004365592 2022-06-25 13:25:15                 NaT   
266092     101004464909 2022-09-02 12:34:00                 NaT   
271889     101004545274 2022-10-31 14:47:37                 NaT   
273937     101004581723 2022-11-25 10:12:40                 NaT   
276291     101004612126 2022-12-28 05:55:00                 NaT   

                 closed_dt   ontime case_status  \
926    2022-02-14 11:42:58   ONTIME      Closed   
3165                   NaT   ONTIME        Open   
12728  2022-05-23 17:37:39   ONTIME      Closed   
19321                  NaT   ONTIME        Open   
19608                  NaT  OVERDUE        Open   
...                    ...      ...         ...   
258838 2022-06-25 13:29:59   ONTIME      Closed   
266092                 NaT   ONTIME        Open   
271889 2022-11-01 06:15:57   ONTIME      Closed   
273937 2022-11-25 10:13:56   ONTIME      Closed   
276291 2023-01-05 16:24:13   ONTIME      Closed   

                                           closure_reason  \
926     Case Closed. Closed date : 2022-02-14 11:42:58...   
3165                                                        
12728   Case Closed. Closed date : 2022-05-23 17:37:39...   
19321                                                       
19608                                                       
...                                                   ...   
258838  Case Closed. Closed date : 2022-06-25 13:29:59...   
266092                                                      
271889  Case Closed. Closed date : 2022-11-01 06:15:57...   
273937  Case Closed. Closed date : 2022-11-25 10:13:56...   
276291  Case Closed. Closed date : 2023-01-05 16:24:13...   

                                   case_title  \
926     Dumpster & Loading Noise Disturbances   
3165                          Fire Department   
12728   Dumpster & Loading Noise Disturbances   
19321              Aircraft Noise Disturbance   
19608                  Valet Parking Problems   
...                                       ...   
258838             Phone Bank Service Inquiry   
266092                      Billing Complaint   
271889                     Bridge Maintenance   
273937             Phone Bank Service Inquiry   
276291  Dumpster & Loading Noise Disturbances   

                                  subject                        reason  ...  \
926                 Neighborhood Services  Neighborhood Services Issues  ...   
3165              Mayor's 24 Hour Hotline               Fire Department  ...   
12728               Neighborhood Services  Neighborhood Services Issues  ...   
19321             Mayor's 24 Hour Hotline                      Massport  ...   
19608   Transportation - Traffic Division                         Valet  ...   
...                                   ...                           ...  ...   
258838  Transportation - Traffic Division   Office of The Parking Clerk  ...   
266092    Boston Water & Sewer Commission                       Billing  ...   
271889            Public Works Department            Bridge Maintenance  ...   
273937  Transportation - Traffic Division   Office of The Parking Clerk  ...   
276291              Neighborhood Services  Neighborhood Services Issues  ...   

       police_district                   neighborhood  \
926                 A1                       Back Bay   
3165               C11                     Dorchester   
12728              D14             Allston / Brighton   
19321              A15                    Charlestown   
19608               A1

This set of records includes all the neighborhood services subject records we just looked at, and the rest of them are mostly open and contain little information. It seems like they might get routed to other departments that don't end up getting back to 311 to have the case closed. These reason categories are:



```
Fire Department                         12
Office of The Parking Clerk              9
Bridge Maintenance                       7
Billing                                  7
Massport                                 6
Valet                                    4
Alert Boston                             3
MBTA                                     1
```

It seems that some of our category values might be obsolete. Since our goal is to predict time to close and whether a case will be closed moving forward, it might be a good idea to look at the currently available android app and see what values are available to the user to select, and which categories might be assigned by the 311 agents after receiving a new case. Let's add this to the to-dos at the end of this notebook.

For now let's keep looking at our minority category values, continuing with any department category value with fewer than 30 records:



In [ ]:
department_filter = df2022['department'].isin(department_counts[department_counts < 30].index)
display(df2022[department_filter])

case_enquiry_id             open_dt           target_dt  \
546        101004114962 2022-01-03 14:21:00 2022-01-04 14:21:27   
10877      101004271503 2022-04-27 12:14:00 2023-04-27 12:14:02   
15393      101004338627 2022-06-12 09:30:00 2022-06-20 08:30:00   
21213      101004421186 2022-08-09 18:02:00                 NaT   
21632      101004426281 2022-08-13 14:40:00                 NaT   
22250      101004444094 2022-08-19 10:24:00 2022-09-18 10:24:56   
27681      101004518642 2022-10-06 15:25:00                 NaT   
33910      101004128101 2022-01-17 09:52:00 2023-01-17 09:52:43   
34205      101004592444 2022-12-06 08:35:00 2022-12-07 08:35:31   
34207      101004592454 2022-12-06 08:38:00 2023-12-06 08:38:41   
34646      101004142910 2022-01-21 12:33:00 2022-02-04 12:33:48   
35613      101004601737 2022-12-15 10:57:00 2022-12-16 10:57:16   
37355      101004605145 2022-12-19 14:26:00 2022-12-20 14:26:01   
42962      101004224983 2022-03-18 09:26:00 2022-03-25 09:26:15   
43723      101004229627 2022-03-22 11:50:00 2022-04-05 11:50:48   
44376      101004233593 2022-03-25 15:07:00 2022-04-01 15:07:00   
45128      101004114265 2022-01-02 20:34:00                 NaT   
48382      101004248336 2022-04-07 14:52:00 2023-04-07 14:52:25   
48943      101004249954 2022-04-08 17:08:00 2022-04-08 17:08:09   
61039      101004194181 2022-02-18 15:45:00 2022-02-21 15:45:27   
71076      101004214156 2022-03-08 13:12:00 2023-03-08 13:12:41   
73588      101004244175 2022-04-04 14:18:00 2023-04-04 14:18:47   
76729      101004357965 2022-06-19 13:02:00 2022-06-21 08:30:00   
88662      101004261401 2022-04-19 12:20:00 2022-04-20 12:21:00   
88744      101004285526 2022-05-09 15:03:00 2023-05-09 15:03:19   
89963      101004279621 2022-05-04 06:44:00 2022-05-05 08:30:00   
134745     101004413943 2022-08-03 17:38:00                 NaT   
138398     101004389834 2022-07-15 12:09:00                 NaT   
146162     101004454937 2022-08-27 21:59:00 2022-09-12 08:30:00   
154055     101004238296 2022-03-30 10:30:00 2022-04-29 10:30:51   
162758     101004310764 2022-05-20 19:38:00 2023-05-20 19:38:04   
169344     101004323943 2022-05-31 17:33:33 2022-06-15 08:30:00   
171042     101004271771 2022-04-27 14:50:00 2022-05-11 14:50:30   
174031     101004277706 2022-05-02 15:38:00 2023-05-02 15:38:00   
203806     101004367267 2022-06-27 11:09:00 2022-06-30 11:09:20   
205973     101004598440 2022-12-12 10:11:00                 NaT   
218050     101004605149 2022-12-19 14:27:00 2023-12-19 14:27:57   
218797     101004420667 2022-08-09 11:50:00 2022-08-23 11:50:50   
219227     101004422434 2022-08-10 15:09:00 2023-08-10 15:09:05   
232627     101004521782 2022-10-10 12:05:00 2022-10-24 12:05:24   
236650     101004606925 2022-12-21 10:43:00 2023-12-21 10:43:07   
240652     101004599789 2022-12-13 12:53:00                 NaT   
242849     101004125124 2022-01-13 12:11:00 2022-01-20 12:11:18   
242870     101004125482 2022-01-13 16:59:00 2022-01-27 16:59:07   
243307     101004131181 2022-01-19 20:33:00 2022-01-21 08:30:00   
254960     101004296705 2022-05-18 07:34:00 2022-05-19 08:30:00   
263318     101004421881 2022-08-10 10:14:00 2022-08-11 10:14:47   
270801     101004532303 2022-10-19 10:32:00 2024-10-08 10:32:29   
271519     101004540868 2022-10-27 10:54:00 2022-11-10 10:54:26   

                 closed_dt   ontime case_status  \
546    2022-05-05 11:27:24  OVERDUE      Closed   
10877                  NaT   ONTIME        Open   
15393                  NaT  OVERDUE        Open   
21213  2022-08-16 11:03:36   ONTIME      Closed   
21632  2022-09-07 09:36:56   ONTIME      Closed   
22250  2022-09-09 09:54:03   ONTIME      Closed   
27681  2022-11-07 15:01:01   ONTIME      Closed   
33910  2022-01-19 16:57:43   ONTIME      Closed   
34205  2022-12-06 14:20:39   ONTIME      Closed   
34207  2022-12-07 10:27:35   ONTIME      Closed   
34646  2022-03-16 13:27:42  OVERDUE      Closed   
35613  2022-12-20 15:05:39  OVERDUE      Clos

These cases appear to fall in three meaningful groups according to the 3 departments involved. This seems to show that department is a very meaningful categorical variable, and the minority values are important as well. These 3 are disability related requests, which we would definitely want to keep, tree maintenance requests, another important category, and requests related to outdoor dining, which were routed to an "ECON" department. We might want to compare a basic model which only uses the department value as a feature to our more complex models as a heuristic for whether additional features actually improve predictions. Let's add this to our to-dos as well. 

To-Dos:

1. look at the currently available android app and see what values are available to the user to select, and which categories might be assigned by the 311 agents after receiving a new case.
2. compare a basic model which only uses the department value as a feature to our more complex models as a heuristic for whether additional features actually improve predictions.